Testing strategy (#432)
aragon/aragon-cli: TESTING.md

# Testing

> We have two different processes. One for running unit and integration tests and another one for functional tests. This is because functional tests usually take longer.

Reviewer: ```suggestion
We have two different processes. One for running unit and integration tests and another one for functional tests (also called end-to-end tests). This is because functional tests usually take longer.
```

> unit tests
> integration tests
> functional tests

Let's add some links or definitions to these concepts, e.g.:

https://en.wikipedia.org/wiki/Unit_testing
https://en.wikipedia.org/wiki/Integration_testing
https://en.wikipedia.org/wiki/Functional_testing

Maybe we could include a picture like this pyramid:
https://medium.com/@Colin_But/define-testing-strategy-using-the-testing-pyramid-1dabee37e823

Unit + integration test are included on each package under test folder on root (e.g. [aragon-cli](https://github.com/aragon/aragon-cli/tree/master/packages/aragon-cli/test/)). Functional test have their own package [`e2e-tests`](https://github.com/aragon/aragon-cli/tree/master/packages/e2e-tests).

We suggest to run unit + integration test on every change, functional tests only before commits.

## Testing environment

The testing environment consist of the following tools:

1. Structure and syntax -> [AVA framework](https://github.com/avajs/ava)
2. Assertion functions -> [AVA assertions](https://github.com/avajs/ava/blob/master/docs/03-assertions.md)
3. Mocks, spies, and stubs -> [Sinon.JS](https://sinonjs.org)
4. Code coverage -> [Istanbul](https://istanbul.js.org) with [nyc](https://github.com/istanbuljs/nyc) wrapper. We also use [Coveralls](https://coveralls.io) for adding coverage reporting to [Travis CI](https://travis-ci.org) flow.

> **Note:** In some instances you might found useful to use [proxyquire](https://github.com/thlorenz/proxyquire) to mock an Aragon wrapper.

Reviewer: I think this applies only to `aragon.js` 🤔, maybe have this instead:
```suggestion
**Note:** In some instances you might found useful to use [proxyquire](https://github.com/thlorenz/proxyquire) to mock dependencies that should not be evaluated.
```

### Running tests

#### Scripts

Each package comes with the following scripts to run unit and integration tests:

```json
{
  "test": "nyc ava",

>   "watch": "ava --watch"

Author: Great, that was my first option but then I changed it cause we had it like that on the `e2e-tests` package. I'll change all the `package.json` files as well.
